### PR TITLE
[SPARK-27770] [SQL]  Port AGGREGATES.sql [Part 1]

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?

This PR is to port AGGREGATES.sql from PostgreSQL regression tests. https://github.com/postgres/postgres/blob/02ddd499322ab6f2f0d58692955dc9633c2150fc/src/test/regress/sql/aggregates.sql#L1-L143

The expected results can be found in the link: https://github.com/postgres/postgres/blob/master/src/test/regress/expected/aggregates.out

When porting the test cases, found three PostgreSQL specific features that do not exist in Spark SQL.
- https://issues.apache.org/jira/browse/SPARK-27765: Type Casts: expression::type
- https://issues.apache.org/jira/browse/SPARK-27766: Data type: POINT(x, y)
- https://issues.apache.org/jira/browse/SPARK-27767: Built-in function: generate_series

Also, found two bugs:
- https://issues.apache.org/jira/browse/SPARK-27768: Infinity, -Infinity, NaN should be recognized in a case insensitive manner
- https://issues.apache.org/jira/browse/SPARK-27769: Handling of sublinks within outer-level aggregates.

This PR also fixes the error message when the column can't be resolved. 

For running the regression tests, this PR also added three tables `aggtest`, `onek` and `tenk1` from the postgreSQL data sets: https://github.com/postgres/postgres/tree/02ddd499322ab6f2f0d58692955dc9633c2150fc/src/test/regress/data

## How was this patch tested?
N/A

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+group by ten
+having exists (select 1 from onek b where sum(distinct a.four) = b.four);
+
+-- this should fail because subquery has an agg of its own in WHERE
+select ten, sum(distinct four) from onek a
+group by ten
+having exists (select 1 from onek b
+               where sum(distinct a.four + b.four) = b.four);
+
+-- [SPARK-27769] Test handling of sublinks within outer-level aggregates.
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: https://issues.apache.org/jira/browse/SPARK-27769

File: sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+select sum(CAST(null AS Decimal(38,0))) from range(1,4);
+select sum(CAST(null AS DOUBLE)) from range(1,4);
+select avg(CAST(null AS int)) from range(1,4);
+select avg(CAST(null AS long)) from range(1,4);
+select avg(CAST(null AS Decimal(38,0))) from range(1,4);
+select avg(CAST(null AS DOUBLE)) from range(1,4);
+select sum(CAST('NaN' AS DOUBLE)) from range(1,4);
+select avg(CAST('NaN' AS DOUBLE)) from range(1,4);
+
+-- [SPARK-27768] verify correct results for infinite inputs
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: https://issues.apache.org/jira/browse/SPARK-27768

**Comment by dongjoon-hyun** [comment]
Great! @gatorsmile . For the `infinity`, it's a Java issue. We support `Infinity` instead of `infinity`.
```
scala> java.lang.Double.parseDouble("Infinity")
res0: Double = Infinity

scala> java.lang.Double.parseDouble("infinity")
java.lang.NumberFormatException: For input string: "infinity"
```

File: sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: I removed the `correctness` label because we support `Infinity` correctly.
```scala
scala> sql("SELECT CAST('Infinity' AS DOUBLE), CAST('infinity' AS DOUBLE)").show
+------------------------+------------------------+
|CAST(Infinity AS DOUBLE)|CAST(infinity AS DOUBLE)|
+------------------------+------------------------+
|                Infinity|                    null|
+------------------------+------------------------+
```

File: sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: I removed the `correctness` label because we support `Infinity` correctly.
```scala
scala> sql("SELECT CAST('Infinity' AS DOUBLE), CAST('infinity' AS DOUBLE)").show
+------------------------+------------------------+
|CAST(Infinity AS DOUBLE)|CAST(infinity AS DOUBLE)|
+------------------------+------------------------+
|                Infinity|                    null|
+------------------------+------------------------+
```

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: Since we already file SPARK-27768, shall we use `Infinity` here and move forward for the test coverage first?

Comment (gatorsmile) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: Thanks for pointing out these. We should parse these constants in a case insensitive way. See the description in https://www.postgresql.org/docs/9.3/datatype-numeric.html

> When writing these values as constants in an SQL command, you must put quotes around them, for example UPDATE table SET x = 'Infinity'. On input, these strings are recognized in a case-insensitive manner.

Comment (gatorsmile) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: Updated the JIRA and make the description more accurate.

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/inputs/aggregates_part1.sql]: Thanks!

**Comment by dongjoon-hyun** [comment]
Could you fix the UT failure?
```
[info] - aggregates_part1.sql *** FAILED *** (3 seconds, 720 milliseconds)
```

**Comment by gatorsmile** [comment]
cc @dongjoon-hyun @cloud-fan @HyukjinKwon @wangyum

**Comment by dilipbiswal** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -109,7 +109,7 @@ trait CheckAnalysis extends PredicateHelper {
 
         operator transformExpressionsUp {
           case a: Attribute if !a.resolved =>
-            val from = operator.inputSet.map(_.qualifiedName).mkString(", ")
+            val from = operator.inputSet.toSeq.map(_.qualifiedName).mkString(", ")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Ur, we should not do this in `[TEST]` PR because this is `sql/catalyst/src/main`. Please revert this and make another PR for this if needed.

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This minor fix is to improve the error messages. It does not worth a separate PR. I removed [TEST]

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: I updated the PR description.

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Also, this is to fix the non-deterministic test failure in aggregate-part1.sql

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Got it. Now, this becomes non `[TEST]` and this fix is described in the PR description. Looks good.
> This PR also fixes the error message when the column can't be resolved.

File: sql/core/src/test/resources/sql-tests/inputs/pgSQL/aggregates_part1.sql [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+select avg(CAST(null AS int)) from range(1,4);
+select avg(CAST(null AS long)) from range(1,4);
+select avg(CAST(null AS Decimal(38,0))) from range(1,4);
+select avg(CAST(null AS DOUBLE)) from range(1,4);
+select sum(CAST('NaN' AS DOUBLE)) from range(1,4);
+select avg(CAST('NaN' AS DOUBLE)) from range(1,4);
+
+-- [SPARK-27768] verify correct results for infinite inputs
+SELECT avg(CAST(x AS DOUBLE)), var_pop(CAST(x AS DOUBLE))
+FROM (VALUES (CAST('1' AS DOUBLE)), (CAST('Infinity' AS DOUBLE))) v(x);
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/pgSQL/aggregates_part1.sql]: Thank you for updating this, too.

File: sql/core/src/test/resources/test-data/postgresql/onek.data [hunk 5]
```diff
@@ -0,0 +1,1000 @@
+147	0	1	3	7	7	7	47	147	147	147	14	15	RFAAAA	AAAAAA	AAAAxx
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/test-data/postgresql/onek.data]: In the PR description, could you describe the location of these data sources, too?

Comment (gatorsmile) [reply sql/core/src/test/resources/test-data/postgresql/onek.data]: done.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
Thanks, Merged to master!
